pkg/trace/writer/newwriter/sender.go

```diff
@@ -0,0 +1,382 @@
+func newSender(cfg *senderConfig) *sender {
+	return &sender{
+		cfg:    cfg,
+		climit: make(chan struct{}, cfg.maxConns),
+		list:   list.New(),
+	}
+}
+
+// Push pushes p onto the sender, to be written to the destination.
+func (q *sender) Push(p *payload) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	// enqueue the payload
+	q.enqueueLocked(p)
+
+	if !q.scheduled {
+		// no flush is scheduled; start one
+		q.scheduled = true
+		q.wg.Add(1)
+		go q.flush()
+	}
+}
+
+// flush drains and sends the entire queue. If anything comes in while flushing
+// or if some of the payloads fail to send as retriable, further follow up flushes
+// are scheduled.
+func (q *sender) flush() {
+	startTime := time.Now()
+
+	// we drain the queue, which is a blocking operation, meaning that
+	// new payloads are stopped from coming in while this happens; but
+	// it is a fast.
+	payloads := q.drainQueue()
+
+	// we send the payloads we've retrieved; while we do this, more payloads
+	// can join the list.
+	done, retries := q.sendPayloads(payloads)
+
+	if done > 0 {
+		q.recordEvent(eventTypeFlushed, &eventData{
+			count:    int(done),
+			duration: time.Since(startTime),
+		})
+	}
+
+	// we reassess the state of the list and check if further flushes need
+	// to be scheduled
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	defer q.wg.Done()
+
+	if q.list.Len() == 0 {
+		// the list is empty; no further flushing needs to be triggered
+		q.attempt = 0
+		q.scheduled = false
+		return
+	}
+	// the list is not empty
+	q.scheduled = true
+	if retries > 0 {
+		// some sends failed as retriable; we need to back off a bit
+		q.attempt++
+		q.wg.Add(1)
+		q.timer = time.AfterFunc(backoffDuration(q.attempt), q.flush)
+		return
+	}
+	// all items in the list are new; flush immediately
+	q.wg.Add(1)
+	go q.flush()
+}
+
+// drainQueue drains the entire queue and returns all the payloads that were in it.
+func (q *sender) drainQueue() []*payload {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+	var payloads []*payload
+	for q.list.Len() > 0 {
+		v := q.list.Remove(q.list.Front())
+		payloads = append(payloads, v.(*payload))
+	}
+	q.size = 0
+	return payloads
+}
+
+// sendPayloads concurrently sends the given list of payloads. It returns
+// the number of payloads that were added back onto the queue to be retried
+// again later due to an error.
+func (q *sender) sendPayloads(payloads []*payload) (done, retries uint64) {
+	var wg sync.WaitGroup
+	for _, p := range payloads {
+		q.climit <- struct{}{}
+		wg.Add(1)
+		go func(p *payload) {
+			defer func() { <-q.climit }()
+			defer wg.Done()
+			req, err := p.httpRequest(q.cfg.url)
+			if err != nil {
+				log.Errorf("http.Request: %s", err)
+				return
+			}
+			start := time.Now()
+			err = q.do(req)
+			stats := &eventData{
+				bytes:    len(p.body),
+				count:    1,
+				duration: time.Since(start),
+				err:      err,
+			}
+			switch err.(type) {
+			case *retriableError:
+				// request failed again, but can be retried
+				q.enqueue(p)
```

Comment:
Maybe we should call a `enqueueOrDrop` in the case of a retry with a full queue so we drop the retry and not the more recent data that was since enqueued.

Reply:
Yep. Great idea. I was pondering around this myself but didn't wanna complicate things. It does make sense. Added it and called it `enqueueAgain`.